src/Features/Core/Portable/AddDebuggerDisplay/AbstractAddDebuggerDisplayCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddImports;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.AddDebuggerDisplay
+{
+    internal abstract class AbstractAddDebuggerDisplayCodeRefactoringProvider<TTypeDeclarationSyntax, TMethodDeclarationSyntax> : CodeRefactoringProvider
+        where TTypeDeclarationSyntax : SyntaxNode
+        where TMethodDeclarationSyntax : SyntaxNode
+    {
+        private static readonly SyntaxAnnotation s_trackingAnnotation = new SyntaxAnnotation();
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var type = await context.TryGetRelevantNodeAsync<TTypeDeclarationSyntax>().ConfigureAwait(false);
+
+            if (type is { })
+            {
+                if (!DeclaresToStringOverride(type)) return;
+            }
+            else
+            {
+                var method = await context.TryGetRelevantNodeAsync<TMethodDeclarationSyntax>().ConfigureAwait(false);
+                if (!IsToStringOverride(method)) return;
+
+                type = method.FirstAncestorOrSelf<TTypeDeclarationSyntax>();
+            }
+
+            if (HasDebuggerDisplayAttribute(type))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(new CodeAction(
+                FeaturesResources.Add_DebuggerDisplay,
+                cancellationToken => ApplyAsync(context.Document, type, cancellationToken)));
+        }
+
+        private async Task<Document> ApplyAsync(Document document, TTypeDeclarationSyntax type, CancellationToken cancellationToken)
+        {
+            var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var newAttribute = generator
+                .Attribute("System.Diagnostics.DebuggerDisplayAttribute", generator.LiteralExpression("{ToString(),nq}"))
```

Comment:
`generator.TypeExpression(compilation.GetTypeByMetadataName(...))`.  Note: also make sure we know about this type in the compilation in the Register phase.

Reply:
@CyrusNajmabadi Why does the `string name` overload exist and why is it the only one with params?

Reply:
@CyrusNajmabadi Why does the `string name` overload exist and why is it the only one with params?